soc: rp2350: fix pinctrl and PWM for GPIOs 32 through 47

tari opened the pull request:
The rpi_pico pinctrl driver only supported up to 32 GPIOs, while RP2350 has up to 48. This changes the driver to support the additional ones, by allocating one more bit in pinctrl values.

The rpi_pico PWM driver also failed to handle PWM channels 8 through 11, which were added on RP2350. That driver is fixed to support up to the required 12 channels, and the pinctrl macros modified to accurately reflect which PWM channels can be connected to which pin.

tari committed:
soc: rp2350: correctly handle GPIOs >31

RP2350 has 48 GPIOs, where only the first 30 are broken out to pins on
RP2350A (same as RP2040) and the remaining 18 are only usable on RP2350B.
This change makes the soc pinctrl driver support GPIOs above 31, where
previously it was impossible to configure GPIOs 32 through 47.

Tested on RP2350B, confirming that GPIO44 can be correctly configured for
PWM.

Signed-off-by: Peter Marheine <peter@taricorp.net>

ajf58 commented on dts/bindings/pwm/raspberrypi,pico-pwm.yaml:
```diff
@@ -87,6 +87,38 @@ properties:
     type: int
     description: See divider-frac-0 for help
 
+  divider-int-8:
```
Might be worth making this more explicit in the devicetree binding, e.g.
```suggestion
# N.b. PWM 8-11 are only valid for the RP2350
  divider-int-8:
```

tari replied:
Better yet, I added information to that effect to the field description so it's available in the documentation.

ajf58 commented on drivers/pwm/pwm_rpi_pico.c:
```diff
@@ -17,8 +17,21 @@ LOG_MODULE_REGISTER(pwm_rpi_pico, CONFIG_PWM_LOG_LEVEL);
+#else
+#define PWM_RPI_PICO_NUM_SLICES 8
+#endif
+BUILD_ASSERT(NUM_PWM_SLICES == PWM_RPI_PICO_NUM_SLICES,
+			 "PWM_RPI_PICO_NUM_SLICES misconfigured for SOC");
```
This is a neat way of dealing with this potential issue. I've added a suggestion to call this out in the devicetree binding as well.

ajf58 reviewed (changes requested):
nit (sorry): in the commit message "available on RP2030" is a minor but confusing typo that might be a bit of a pain when searching the commit history in the future.

Other than that this looks great.

tari commented:
> in the commit message "available on RP2030" is a minor but confusing typo that might be a bit of a pain when searching the commit history in the future.

I think both instances of that accurately reflect what I meant to write; what wording do you have in mind?

tari commented:
> > in the commit message "available on RP2030" is a minor but confusing typo that might be a bit of a pain when searching the commit history in the future.
> 
> I think both instances of that accurately reflect what I meant to write; what wording do you have in mind?

Never mind, I realized it should be RP2040.

ThreeEights commented on include/zephyr/dt-bindings/pinctrl/rpi-pico-rp2350b-pinctrl.h:
```diff
@@ -79,14 +79,14 @@
-#define PWM_14B_P45 RP2XXX_PINMUX(45, RP2_PINCTRL_GPIO_FUNC_PWM)
-#define PWM_15A_P46 RP2XXX_PINMUX(46, RP2_PINCTRL_GPIO_FUNC_PWM)
-#define PWM_15B_P47 RP2XXX_PINMUX(47, RP2_PINCTRL_GPIO_FUNC_PWM)
+#define PWM_8A_P40 RP2XXX_PINMUX(40, RP2_PINCTRL_GPIO_FUNC_PWM)
```
Good fix!

tari committed:
soc: rp2350: support PWM channels >8

RP2350 adds four more PWM slices from the eight available on RP2040,
which are only broken out to package pins on RP2350B. This change fixes
the driver to support the correct number of slices on RP2350.

Tested by confirming that PWM can correctly be configured on GPIO 44 of
RP2350B.

Signed-off-by: Peter Marheine <peter@taricorp.net>

tari committed:
dt-bindings: rp2350: correct PWM names for GPIOs 40..47

The macros for PWM output on these GPIOs had names that referred to PWM
channels that do no exist on this chip- it only has 12 channels, where
channels 8..11 are exposed on both GPIOs 32..39 and 40..47.

Signed-off-by: Peter Marheine <peter@taricorp.net>
